DataMigration/PostgresDB/DataProvider.cs

```diff
@@ -0,0 +1,179 @@
+                                    dataToInsert.StatusId,
+                                    string.IsNullOrEmpty(dataToInsert.ErrorMessage) ? "null" : dataToInsert.ErrorMessage,
+                                    NormalizeJsonData(dataToInsert.Data),
+                                    dataToInsert.CreatedAt,
+                                    dataToInsert.UpdatedAt);
+
+
+            int affected;
+            using (var command = new NpgsqlCommand(commandText, GetOpenedConnection()))
+            {
+                affected = command.ExecuteNonQuery();
+            }
+
+            LogEventHappened?.Invoke($"Data is inserted! Success! Row(s) affected: {affected}", LogLevel.Debug);
+            return affected;
+        }
+
+        public int InsertHistoricalOcrData(IList<HistoricalOcrData> dataToInsert)
+        {
+            LogEventHappened?.Invoke("Inserting data...", LogLevel.Debug);
+            string commandText = GetInsertQueryString(dataToInsert);
+
+            int affected;
+            using (var command = new NpgsqlCommand(commandText, GetOpenedConnection()))
+            {
+                affected = command.ExecuteNonQuery();
+            }
+
+            LogEventHappened?.Invoke($"Data is inserted! Success! Row(s) affected: {affected}", LogLevel.Debug);
+            return affected;
+        }
+
+        private int DeleteData(IList<HistoricalOcrData> data)
+        {
+            LogEventHappened?.Invoke("Deleting selected data...", LogLevel.Debug);
+            string query = GetQueryStringToRemoveData(data);
+            int affected;
+            using (var command = new NpgsqlCommand(query, GetOpenedConnection()))
+            {
+                affected = command.ExecuteNonQuery();
+                LogEventHappened?.Invoke($"Deleted! Affected row(s): {affected}", LogLevel.Debug);
+            }
+            return affected;
+        }
+
+        private string GetInsertQueryString(IList<HistoricalOcrData> dataToInsert)
+        {
+            string commandText = "INSERT INTO historical_ocr_data VALUES ";
+            StringBuilder sb = new StringBuilder(commandText);
+
+            for (int i = 0; i < dataToInsert.Count - 1; i++)
+            {
+                sb.Append(string.Format("({0}, '{1}', {2}, {3}, '{4}', '{5}', '{6}'), ",
+                                    dataToInsert[i].TenantId,
+                                    dataToInsert[i].FullFilePath,
+                                    dataToInsert[i].StatusId,
+                                    string.IsNullOrEmpty(dataToInsert[i].ErrorMessage) ? "null" : dataToInsert[i].ErrorMessage,
+                                    NormalizeJsonData(dataToInsert[i].Data),
+                                    dataToInsert[i].CreatedAt,
+                                    dataToInsert[i].UpdatedAt));
+            }
+
+            int lastIndex = dataToInsert.Count - 1;
+            sb.Append(string.Format("({0}, '{1}', {2}, {3}, '{4}', '{5}', '{6}')",
```

Comment:
why are you using that duplication instead of using one more iteration in the loop?